src/emqx_client.erl

```diff
@@ -248,12 +283,63 @@ parse_subopt([{nl, false} | Opts], Result) ->
 parse_subopt([{qos, QoS} | Opts], Result) ->
     parse_subopt(Opts, Result#{qos := ?QOS_I(QoS)}).
 
+unify_qos(Opts) -> lists:map(fun({qos, QoS}) -> {qos, ?QOS_I(QoS)}; (X) -> X end, Opts).
+
+-spec(request(client(), topic(), payload(), qos() | [pubopt()])
+        -> ok | {ok, packet_id()} | {error, term()}).
+request(Client, Topic, Payload, QoS) when ?IS_QOS(QoS);
+                                          ?IS_QOS_NAME(QoS) ->
+    request(Client, Topic, Payload, [{qos, QoS}]);
+request(Client, Topic, Payload, Opts0) ->
+    request(Client, Topic, #{}, Payload, unify_qos(Opts0)).
+
+%% request(Client, Topic, Payload, QoS) when is_binary(Topic), is_atom(QoS) ->
+%%     request(Client, Topic, Payload, [{qos, ?QOS_I(QoS)}]);
+%% request(Client, Topic, Payload, QoS) when is_binary(Topic), ?IS_QOS(QoS) ->
+%%     request(Client, Topic, Payload, [{qos, QoS}]);
+%% request(Client, Topic, Payload, Opts) when is_binary(Topic), is_list(Opts) ->
+%%     request(Client, Topic, #{}, Payload, Opts).
+
+-spec(request(client(), topic(), properties(), payload(), [pubopt()])
+        -> ok | {ok, packet_id()} | {error, term()}).
+request(Client, Topic, Properties, Payload, Opts)
+    when is_binary(Topic), is_map(Properties), is_list(Opts) ->
+    ok = emqx_mqtt_props:validate(Properties),
+    Retain = proplists:get_bool(retain, Opts),
+    Group = proplists:get_value(group, Opts, <<>>),
+    TimeOut = proplists:get_value(timeout, Opts, ?RESPONSE_TIMEOUT_SECONDS),
+    QoS = ?QOS_I(proplists:get_value(qos, Opts, ?QOS_0)),
+    {ok, ResponseTopic} = sub_response_topic(Client, QoS, {Group, Topic}),
+    ClientId = gen_statem:call(Client, client_id),
+    NewProperties = maps:merge(Properties, #{'Response-Topic' => ResponseTopic,
+                                             'Correlation-Data' => ClientId}),
+    publish(Client, #mqtt_msg{qos  = QoS,
+                              retain = Retain,
+                              topic = ResponseTopic,
+                              props = NewProperties,
+                              payload = iolist_to_binary(Payload)}),
+    receive_response(Client, ClientId, TimeOut).
+
+receive_response(Client, ClientId, TimeOut) ->
+    receive
+        {publish, Response} ->
+            {ok, Properties} = maps:find(properties, Response),
+            case maps:find('Correlation-Data', Properties) of
+                {ok, ClientId} ->
+                    maps:find(payload, Response);
+                _ ->
+                    receive_response(Client, ClientId, TimeOut)
```

Comment:
this recursion will rest the timeout value for `after Timeout`.
e.g. if timeout is 10 seconds, and it spent 9 seconds to receive a message, but turns out isn't not my correlation-data. so it decide to discard and keep waiting, then the next message come after another 9 seconds again.
in such case, the total time spent on this function is 18 seconds.

A more accurate way is perhaps to call `Ref = erlang:start_timer(Timeout, self(), response)`, then receive `{timeout, Ref, response}` in the receive clause.